### PR TITLE
Implement StrategyStargateStakingTridentRouter

### PR DESCRIPTION
**Pull request by andreb0x**
the existing stargate strategy routes swaps through `swapExactTokensForTokens`. however, in polygon STG liquidity is provided in sushi's trident router; the trident router does not define a `swapExactTokensForTokens` method. this strategy simply modifies that strategies swaps to use trident's `exactInput` method

trident router implementation: https://github.com/sushiswap/trident/blob/master/contracts/TridentRouter.sol

### CODE REVIEW

**Comment by andreb0x** [comment]
current diff to StrategyStargateStaking:
https://www.diffchecker.com/cbG3fbCd

**Comment by andreb0x** [comment]
found implementation of pool contract https://github.com/sushiswap/trident/blob/master/contracts/pool/constant-product/ConstantProductPool.sol use this to setup data arg passed to pool .swap method

**Comment by andreb0x** [comment]
https://www.diffchecker.com/FnpIeVJq

**Comment by andreb0x** [comment]
currently breaking with: 

Screen Shot 2022-04-07 at 6 26 58 PM

**Comment by andreb0x** [comment]
currently breaking with 
Screen Shot 2022-04-07 at 6 37 05 PM

**Comment by andreb0x** [comment]
successful compile. pending tests

**Comment by andreb0x** [comment]
when deploying strat:
`ProviderError: Error: VM Exception while processing transaction: invalid opcode`

**Comment by andreb0x** [comment]
https://www.diffchecker.com/WJf7Vsrm

File: contracts/BIFI/strategies/Stargate/StrategyStargateStakingTridentRouter.sol [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    ITridentRouter.Path[] public outputToLp0Path;
+
+    event StratHarvest(address indexed harvester, uint256 wantHarvested, uint256 tvl);
+    event Deposit(uint256 tvl);
+    event Withdraw(uint256 tvl);
+    event ChargedFees(uint256 callFees, uint256 beefyFees, uint256 strategistFees);
+
+    constructor(
+        address _want,
+        uint256[] memory _poolIdAndRouterPoolId,
```
Comment (sircryptomaxi) [inline contracts/BIFI/strategies/Stargate/StrategyStargateStakingTridentRouter.sol:55]: Guessing you ran into the local variable limit here?

Comment (andreb0x) [reply contracts/BIFI/strategies/Stargate/StrategyStargateStakingTridentRouter.sol:55]: correct! is this way of packing inputs into arrays appropriate? or would you prefer it different?

File: contracts/BIFI/strategies/Stargate/StrategyStargateStakingTridentRouter.sol [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+        uint256 strategistFeeAmount = nativeBal.mul(STRATEGIST_FEE).div(MAX_FEE);
+        IERC20(native).safeTransfer(strategist, strategistFeeAmount);
+
+        emit ChargedFees(callFeeAmount, beefyFeeAmount, strategistFeeAmount);
+    }
+
+    // swap tokens 
+    // @dev Ensure pools are tristed before calling this function
+    function tridentSwap(address _tokenIn, uint256 _amountIn, ITridentRouter.Path[] memory _path) internal returns (uint256) {
```
[the remark this answers is not in the document]
Comment (andreb0x) [reply contracts/BIFI/strategies/Stargate/StrategyStargateStakingTridentRouter.sol]: removed

**Comment by andreb0x** [comment]
> I think it's good, would like to see it in action. One thing to do to improve is to collect all trident related functions and abstract it into an inherited contract, that way we keep the strategy clean and specific to the strategy's fundamental functions. Beethoven strats will also need this refactor.

@sircryptomaxi refactored

**Comment by andreb0x** [comment]
Screen Shot 2022-04-21 at 10 29 56 AM
NB failed yarn test not specific to this PR

**Comment by andreb0x** [comment]
deployed strat failing harvest because due to lack of bento box approval. need to call `setMasterContractApproval` with `masterContract = 0xc5017be80b4446988e8686168396289a9a62668e` (trident router) and `from = 0x67f3d7676768eeffbe102036143f06fb385180f7` (strat address). 

next step:  `setMasterContractApproval` requires encoding signature variables `(uint8 v, bytes32 r, bytes32 s)` manually.

**Comment by andreb0x** [comment]
currently failing on transfer; current implementation attempts to transfer from user to pool, but I think bento requires funds to be swapped into bento shares instead

**Pull request closed without merging**